File: NNN-Solution/NNNCSharp/Components/Autodiff/TensorActivations.cs
using NNNCSharp.Components.Utilities;
using System.Numerics;
using System.Runtime.InteropServices;

namespace NNNCSharp.Components.Autodiff;

public partial class Tensor
{
    // Activation functions

    /// <summary>
    /// Applies the Rectified Linear Unit function to the input tensor.
    /// </summary>
    /// <param name="t">Tensor to apply ReLU function to.</param>
    /// <returns>Result tensor of applying ReLU function to the input tensor.</returns>
    public static Tensor ReLU(Tensor t) // ReLU(t) = t (t > 0), 0 (t <= 0)
    {
        Tensor result = GetResultTensor(t, t.Dimensions, t.RequiresGrad);

        // Vectorize inputs and results
        var tVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Data.AsSpan());
        var vzero = Vector<double>.Zero; // preallocate 0's vector
        var rVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Data.AsSpan());

        // Calculate vectorized ReLU result
        for (int i = 0; i < tVecs.Length; i++)
        {
            rVecs[i] = Vector.Max(tVecs[i], vzero);
        }

        // Clean up unvectorized tail
        for (int i = tVecs.Length * VectorSize; i < result.ElementCount; i++)
        {
            result[i] = Math.Max(t[i], 0.0);
        }

        // Connect result tensor to autograd graph if needed
        if (!Inference)
        {
            result._parents.Add(t);

            // Gradient calculation function for r = ReLU(t) -> dr/dt = 1 (t > 0), 0 (t <= 0)
            result._backward = () =>
            {
                if (!t.RequiresGrad) return;

                // Vectorize inputs and gradients
                var tvVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Data.AsSpan());
                var tgVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Grad.AsSpan());
                var vzero = Vector<double>.Zero; // preallocate 0's vector
                var rgVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Grad.AsSpan());

                // Calculate vectorized gradients of parent
                for (int i = 0; i < rgVecs.Length; i++)
                {
                    var mask = Vector.GreaterThan(tvVecs[i], vzero);
                    tgVecs[i] += Vector.ConditionalSelect(mask, rgVecs[i], vzero);
                }

                // Clean up unvectorized tail
                for (int i = rgVecs.Length * VectorSize; i < result.ElementCount; i++)
                {
                    t.Grad[i] += (t[i] > 0.0 ? 1.0 : 0.0) * result.Grad[i];
                }
            };
        }

        return result;
    }

    /// <summary>
    /// Applies the Leaky Rectified Linear Unit function to the input tensor.
    /// </summary>
    /// <param name="t">Tensor to apply Leaky ReLU function to.</param>
    /// <param name="tau">Tau coefficient to use.</param>
    /// <returns>Result tensor of applying Leaky ReLU function to the input tensor.</returns>
    public static Tensor LeakyReLU(Tensor t, double tau) // Leaky ReLU(t) = t (t > 0), Tau * t (t <= 0)
    {
        Tensor result = GetResultTensor(t, t.Dimensions, t.RequiresGrad);

        // Vectorize inputs and results
        var tVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Data.AsSpan());
        var vzero = Vector<double>.Zero; // preallocate 0's vector
        var vtau = new Vector<double>(tau); // splat scalar into vector
        var rVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Data.AsSpan());

        // Calculate vectorized Leaky ReLU result
        for (int i = 0; i < tVecs.Length; i++)
        {
            var mask = Vector.GreaterThan(tVecs[i], vzero);
            rVecs[i] = Vector.ConditionalSelect(mask, tVecs[i], vtau * tVecs[i]);
        }

        // Clean up unvectorized tail
        for (int i = tVecs.Length * VectorSize; i < result.ElementCount; i++)
        {
            result[i] = t[i] > 0.0 ? t[i] : tau * t[i];
        }

        // Connect result tensor to autograd graph if needed
        if (!Inference)
        {
            result._parents.Add(t);

            // Gradient calculation function for r = Leaky ReLU(t) -> dr/dt = 1 (t > 0), Tau (t <= 0)
            result._backward = () =>
            {
                if (!t.RequiresGrad) return;

                // Vectorize inputs and gradients
                var tvVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Data.AsSpan());
                var tgVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Grad.AsSpan());
                var vzero = Vector<double>.Zero; // preallocate 0's vector
                var vone = Vector<double>.One; // preallocate 1's vector
                var vtau = new Vector<double>(tau); // splat scalar into vector
                var rgVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Grad.AsSpan());

                // Calculate vectorized gradients of parent
                for (int i = 0; i < rgVecs.Length; i++)
                {
                    var mask = Vector.GreaterThan(tvVecs[i], vzero);
                    tgVecs[i] += Vector.ConditionalSelect(mask, vone, vtau) * rgVecs[i];
                }

                // Clean up unvectorized tail
                for (int i = rgVecs.Length * VectorSize; i < result.ElementCount; i++)
                {
                    t.Grad[i] += (t[i] > 0.0 ? 1.0 : tau) * result.Grad[i];
                }
            };
        }

        return result;
    }

    /// <summary>
    /// Applies the sigmoid activation function to the input tensor.
    /// </summary>
    /// <param name="t">Tensor to apply sigmoid function to.</param>
    /// <returns>Result tensor of applying sigmoid function to the input tensor.</returns>
    public static Tensor Sigmoid(Tensor t) // Sigmoid(t) = 1 / (1 + e^(-t))
    {
        Tensor result = GetResultTensor(t, t.Dimensions, t.RequiresGrad);

        // Vectorize inputs and results
        var tVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Data.AsSpan());
        var vone = Vector<double>.One; // preallocate 1's vector
        var rVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Data.AsSpan());

        // Calculate vectorized sigmoid result
        for (int i = 0; i < tVecs.Length; i++)
        {
            rVecs[i] = vone / (vone + Vector.Exp(-tVecs[i]));
        }

        // Clean up unvectorized tail
        for (int i = tVecs.Length * VectorSize; i < result.ElementCount; i++)
        {
            result[i] = MathUtils.Sigmoid(t[i]);
        }

        // Connect result tensor to autograd graph if needed
        if (!Inference)
        {
            result._parents.Add(t);

            // Gradient calculation function for r = Sigmoid(t) -> dr/dt = Sigmoid(t) * (1 - Sigmoid(t))
            result._backward = () =>
            {
                if (!t.RequiresGrad) return;

                // Vectorize inputs, results, and gradients
                var tgVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Grad.AsSpan());
                var vone = Vector<double>.One; // preallocate 1's vector
                var rvVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Data.AsSpan());
                var rgVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Grad.AsSpan());

                // Calculate vectorized gradients for parent
                for (int i = 0; i < rgVecs.Length; i++)
                {
                    // Reuse forward results as Sigmoid(t)
                    tgVecs[i] += rvVecs[i] * (vone - rvVecs[i]) * rgVecs[i];
                }

                // Clean up unvectorized tail
                for (int i = rgVecs.Length * VectorSize; i < result.ElementCount; i++)
                {
                    // Reuse forward results as Sigmoid(t)
                    t.Grad[i] += result[i] * (1.0 - result[i]) * result.Grad[i];
                }
            };
        }

        return result;
    }

    /// <summary>
    /// Applies the hyperbolic tangent activation function to the input tensor.
    /// </summary>
    /// <param name="t">Tensor to apply Tanh function to.</param>
    /// <returns>Result tensor of applying Tanh function to the input tensor.</returns>
    public static Tensor Tanh(Tensor t) // Tanh(t) = (e^2t - 1) / (e^2t + 1)
    {
        Tensor result = GetResultTensor(t, t.Dimensions, t.RequiresGrad);

        // Vectorize inputs and results
        var tVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Data.AsSpan());
        var vone = Vector<double>.One; // preallocate 1's vector
        var vtwo = new Vector<double>(2.0); // preallocate 2's vector
        var rVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Data.AsSpan());

        // Calculate vectoried Tanh result
        for (int i = 0; i < tVecs.Length; i++)
        {
            var e2t = Vector.Exp(vtwo * tVecs[i]);
            rVecs[i] = (e2t - vone) / (e2t + vone);
        }

        // Clean up unvectorized tail
        for (int i = tVecs.Length * VectorSize; i < result.ElementCount; i++)
        {
            result[i] = MathUtils.Tanh(t[i]);
        }

        // Connect result tensor to autograd graph if needed
        if (!Inference)
        {
            result._parents.Add(t);

            // Gradient calculation function for r = Tanh(t) -> dr/dt = 1 - Tanh^2(t)
            result._backward = () =>
            {
                if (!t.RequiresGrad) return;

                // Vectorize inputs, results, and gradients
                var tgVecs = MemoryMarshal.Cast<double, Vector<double>>(t.Grad.AsSpan());
                var vone = Vector<double>.One; // preallocate 1's tensor
                var rvVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Data.AsSpan());
                var rgVecs = MemoryMarshal.Cast<double, Vector<double>>(result.Grad.AsSpan());

                // Calculate vectorized gradients of parent
                for (int i = 0; i < rgVecs.Length; i++)
                {
                    // Reuse forward results as Tanh(t)
                    tgVecs[i] += (vone - (rvVecs[i] * rvVecs[i])) * rgVecs[i];
                }

                // Clean up unvectorized tail
                for (int i = rgVecs.Length * VectorSize; i < result.ElementCount; i++)
                {
                    // Reuse forward results as Tanh(t)
                    t.Grad[i] += (1.0 - (result[i] * result[i])) * result.Grad[i];
                }
            };
        }

        return result;
    }

    /// <summary>
    /// Applies the softmax activation function to the input tensor.
    /// </summary>
    /// <param name="t">Tensor to apply softmax function to.</param>
    /// <returns>Result tensor of applying softmax function to the input tensor.</returns>
    public static Tensor Softmax(Tensor t) // Softmax(z_i) = e^z_i / sum(j = 1 to n) [e^z_j]
    {
        Tensor result = GetResultTensor(t, t.Dimensions, t.RequiresGrad);

        int classes = t.Dimensions[^1]; // number of classification probabilities per output
        int batchSize = t.ElementCount / classes; // number of batches of separate outputs

        // Apply softmax function to each batch
        for (int b = 0; b < batchSize; b++)
        {
            int offset = b * classes; // linear offset of current batch

            // Vectorize inputs and results
            var tSlice = t.Data.AsSpan(offset, classes);
            var tVecs = MemoryMarshal.Cast<double, Vector<double>>(tSlice);
            var rSlice = result.Data.AsSpan(offset, classes);
            var rVecs = MemoryMarshal.Cast<double, Vector<double>>(rSlice);

            // Find vectorized maximum value in batch
            var vmax = new Vector<double>(double.MinValue);
            for (int i = 0; i < tVecs.Length; i++)
            {
                vmax = Vector.Max(vmax, tVecs[i]);
            }

            // Find single maximum value in vectorized maximum
            double max = double.MinValue;
            for (int lane = 0; lane < VectorSize; lane++)
            {
                max = Math.Max(max, vmax[lane]);
            }

            // Clean up unvectorized tail
            for (int i = tVecs.Length * VectorSize; i < classes; i++)
            {
                max = Math.Max(max, tSlice[i]);
            }

            // Calculate vectorized exponentiated values and exponentiated sum
            var vmaxSplat = new Vector<double>(max); // splat scalar into vector
            var acc = Vector<double>.Zero; // sum vector
            for (int i = 0; i < tVecs.Length; i++)
            {
                rVecs[i] = Vector.Exp(tVecs[i] - vmaxSplat);
                acc += rVecs[i];
            }
            double sum = Vector.Sum(acc);

            // Clean up unvectorized tail
            for (int i = tVecs.Length * VectorSize; i < classes; i++)
            {
                rSlice[i] = Math.Exp(tSlice[i] - max);
                sum += rSlice[i];
            }

            // Calculate vectorized normalized values
            var vsumSplat = new Vector<double>(sum); // splat scalar into vector
            for (int i = 0; i < rVecs.Length; i++)
            {
                rVecs[i] /= vsumSplat;
            }

            // Clean up unvectorized tail
            for (int i = rVecs.Length * VectorSize; i < classes; i++)
            {
                rSlice[i] /= sum;
            }
        }

        // Connect result tensor to autograd graph if needed
        if (!Inference)
        {
            result._parents.Add(t);

            // Gradient calculation function for r = softmax(t) -> grad_t_i = r_i * (grad_r_i - sum_j(grad_r_j * r_j)) -> Vector-Jacobian product
            result._backward = () =>
            {
                if (!t.RequiresGrad) return;

                // Calculate gradient per batch
                for (int b = 0; b < batchSize; b++)
                {
                    int offset = b * classes; // linear offset of current batch

                    // Vectorize inputs, results, and gradients
                    var tgSlice = t.Grad.AsSpan(offset, classes);
                    var tgVecs = MemoryMarshal.Cast<double, Vector<double>>(tgSlice);
                    var rvSlice = result.Data.AsSpan(offset, classes);
                    var rvVecs = MemoryMarshal.Cast<double, Vector<double>>(rvSlice);
                    var rgSlice = result.Grad.AsSpan(offset, classes);
                    var rgVecs = MemoryMarshal.Cast<double, Vector<double>>(rgSlice);

                    // Calculate vectorized dot product of result gradient and forward softmax result -> sum_j(grad_r_j * r_j)
                    var vdot = Vector<double>.Zero; // product/sum vector
                    for (int i = 0; i < rvVecs.Length; i++)
                    {
                        vdot += rgVecs[i] * rvVecs[i];
                    }
                    double dot = Vector.Sum(vdot);

                    // Clean up unvectorized tail
                    for (int i = rvVecs.Length * VectorSize; i < classes; i++)
                    {
                        dot += rgSlice[i] * rvSlice[i];
                    }

                    // Calculate vectorized gradients for parent
                    var vdotSplat = new Vector<double>(dot); // splat scalar into vector
                    for (int i = 0; i < tgVecs.Length; i++)
                    {
                        tgVecs[i] += rvVecs[i] * (rgVecs[i] - vdotSplat);
                    }

                    // Clean up unvectorized tail
                    for (int i = tgVecs.Length * VectorSize; i < classes; i++)
                    {
                        tgSlice[i] += rvSlice[i] * (rgSlice[i] - dot);
                    }
                }
            };
        }

        return result;
    }
}

File: NNN-Solution/NNN/TensorActivations.cpp
#include "pch.h"
#include "Tensor.h"
#include "MathUtils.h"

/* Activation functions - autograd graph connected */

// Applies the Rectified Linear Unit function to a tensor.
std::shared_ptr<Tensor> Tensor::relu(const std::shared_ptr<Tensor>& t)
{
	auto result = get_result_tensor(t, t->_dimensions, t->requires_grad);

	// Apply ReLU -> r = max(t, 0) = t (t > 0); 0 (t <= 0)
	MathUtils::vector_max(t->_data, 0.0f, result->_data);

	// Connect result tensor to autograd graph if needed
	if (!inference)
	{
		result->_parents.push_back(t);

		// Gradient calculation function -> dr/dt = 1 (t > 0); 0 (t <= 0)
		result->_backward = [t, result]()
			{
				if (!t->requires_grad) return;

				const size_t n = result->element_count();

				const float* const __restrict p_tv = t->_data.data();
				float* const __restrict p_tg = t->_grad.data();
				const float* const __restrict p_rg = result->_grad.data();
				const __m256 reg_0 = _mm256_setzero_ps();

				// AVX2 SIMD vectorize gradient calculation
				size_t i = 0;
				for (; i + 32 <= n; i += 32)
				{
					_mm256_store_ps(&p_tg[i], _mm256_add_ps(_mm256_load_ps(&p_tg[i]), _mm256_blendv_ps(reg_0, _mm256_load_ps(&p_rg[i]), _mm256_cmp_ps(_mm256_load_ps(&p_tv[i]), reg_0, _CMP_GT_OS))));
					_mm256_store_ps(&p_tg[i + 8], _mm256_add_ps(_mm256_load_ps(&p_tg[i + 8]), _mm256_blendv_ps(reg_0, _mm256_load_ps(&p_rg[i + 8]), _mm256_cmp_ps(_mm256_load_ps(&p_tv[i + 8]), reg_0, _CMP_GT_OS))));
					_mm256_store_ps(&p_tg[i + 16], _mm256_add_ps(_mm256_load_ps(&p_tg[i + 16]), _mm256_blendv_ps(reg_0, _mm256_load_ps(&p_rg[i + 16]), _mm256_cmp_ps(_mm256_load_ps(&p_tv[i + 16]), reg_0, _CMP_GT_OS))));
					_mm256_store_ps(&p_tg[i + 24], _mm256_add_ps(_mm256_load_ps(&p_tg[i + 24]), _mm256_blendv_ps(reg_0, _mm256_load_ps(&p_rg[i + 24]), _mm256_cmp_ps(_mm256_load_ps(&p_tv[i + 24]), reg_0, _CMP_GT_OS))));
				}

				for (; i + 8 <= n; i += 8)
				{
					_mm256_store_ps(&p_tg[i], _mm256_add_ps(_mm256_load_ps(&p_tg[i]), _mm256_blendv_ps(reg_0, _mm256_load_ps(&p_rg[i]), _mm256_cmp_ps(_mm256_load_ps(&p_tv[i]), reg_0, _CMP_GT_OS))));
				}

				for (; i < n; ++i)
				{
					if (p_tv[i] > 0.0f)
					{
						p_tg[i] += p_rg[i];
					}
				}
			};
	}

	return result;
}

std::shared_ptr<Tensor> Tensor::leaky_relu(const std::shared_ptr<Tensor>& t, float tau)
{
	auto result = get_result_tensor(t, t->_dimensions, t->requires_grad);

	result->_data = t->_data;

	// Apply Leaky ReLU -> r = t (t > 0); tau * t (t <= 0)

	const size_t n = result->element_count();

	const __m256 reg_tau = _mm256_set1_ps(tau);
	float* const __restrict p_r = result->_data.data();
	const __m256 reg_0 = _mm256_setzero_ps();

	// AVX2 SIMD vectorize result calculation
	size_t i = 0;
	for (; i + 32 <= n; i += 32)
	{
		_mm256_store_ps(&p_r[i], _mm256_blendv_ps(_mm256_load_ps(&p_r[i]), _mm256_mul_ps(_mm256_load_ps(&p_r[i]), reg_tau), _mm256_cmp_ps(_mm256_load_ps(&p_r[i]), reg_0, _CMP_LE_OS)));
		_mm256_store_ps(&p_r[i + 8], _mm256_blendv_ps(_mm256_load_ps(&p_r[i + 8]), _mm256_mul_ps(_mm256_load_ps(&p_r[i + 8]), reg_tau), _mm256_cmp_ps(_mm256_load_ps(&p_r[i + 8]), reg_0, _CMP_LE_OS)));
		_mm256_store_ps(&p_r[i + 16], _mm256_blendv_ps(_mm256_load_ps(&p_r[i + 16]), _mm256_mul_ps(_mm256_load_ps(&p_r[i + 16]), reg_tau), _mm256_cmp_ps(_mm256_load_ps(&p_r[i + 16]), reg_0, _CMP_LE_OS)));
		_mm256_store_ps(&p_r[i + 24], _mm256_blendv_ps(_mm256_load_ps(&p_r[i + 24]), _mm256_mul_ps(_mm256_load_ps(&p_r[i + 24]), reg_tau), _mm256_cmp_ps(_mm256_load_ps(&p_r[i + 24]), reg_0, _CMP_LE_OS)));
	}

	for (; i + 8 <= n; i += 8)
	{
		_mm256_store_ps(&p_r[i], _mm256_blendv_ps(_mm256_load_ps(&p_r[i]), _mm256_mul_ps(_mm256_load_ps(&p_r[i]), reg_tau), _mm256_cmp_ps(_mm256_load_ps(&p_r[i]), reg_0, _CMP_LE_OS)));
	}
	
	for (; i < n; ++i)
	{
		if (p_r[i] <= 0.0f)
		{
			p_r[i] *= tau;
		}
	}

	// Connect result tensor to autograd graph if needed
	if (!inference)
	{
		result->_parents.push_back(t);

		// Gradient calculation function -> dr/dt = 1 (t > 0); tau (t <= 0)
		result->_backward = [t, tau, result, n]()
			{
				const float* const __restrict p_tv = t->_data.data();
				float* const __restrict p_tg = t->_grad.data();
				const __m256 reg_tau = _mm256_set1_ps(tau);
				const float* const __restrict p_rg = result->_grad.data();
				const __m256 reg_0 = _mm256_setzero_ps();
				const __m256 reg_1 = _mm256_set1_ps(1.0f);

				// AVX2 SIMD vectorize gradient calculation
				size_t i = 0;
				for (; i + 32 <= n; i += 32)
				{
					_mm256_store_ps(&p_tg[i], _mm256_fmadd_ps(_mm256_blendv_ps(reg_tau, reg_1, _mm256_cmp_ps(_mm256_load_ps(&p_tv[i]), reg_0, _CMP_GT_OS)), _mm256_load_ps(&p_rg[i]), _mm256_load_ps(&p_tg[i])));
					_mm256_store_ps(&p_tg[i + 8], _mm256_fmadd_ps(_mm256_blendv_ps(reg_tau, reg_1, _mm256_cmp_ps(_mm256_load_ps(&p_tv[i + 8]), reg_0, _CMP_GT_OS)), _mm256_load_ps(&p_rg[i + 8]), _mm256_load_ps(&p_tg[i + 8])));
					_mm256_store_ps(&p_tg[i + 16], _mm256_fmadd_ps(_mm256_blendv_ps(reg_tau, reg_1, _mm256_cmp_ps(_mm256_load_ps(&p_tv[i + 16]), reg_0, _CMP_GT_OS)), _mm256_load_ps(&p_rg[i + 16]), _mm256_load_ps(&p_tg[i + 16])));
					_mm256_store_ps(&p_tg[i + 24], _mm256_fmadd_ps(_mm256_blendv_ps(reg_tau, reg_1, _mm256_cmp_ps(_mm256_load_ps(&p_tv[i + 24]), reg_0, _CMP_GT_OS)), _mm256_load_ps(&p_rg[i + 24]), _mm256_load_ps(&p_tg[i + 24])));
				}

				for (; i + 8 <= n; i += 8)
				{
					_mm256_store_ps(&p_tg[i], _mm256_fmadd_ps(_mm256_blendv_ps(reg_tau, reg_1, _mm256_cmp_ps(_mm256_load_ps(&p_tv[i]), reg_0, _CMP_GT_OS)), _mm256_load_ps(&p_rg[i]), _mm256_load_ps(&p_tg[i])));
				}

				for (; i < n; ++i)
				{
					p_tg[i] += p_tv[i] > 0.0f ? p_rg[i] : tau * p_rg[i];
				}
			};
	}

	return result;
}

std::shared_ptr<Tensor> Tensor::sigmoid(const std::shared_ptr<Tensor>& t)
{
	auto result = get_result_tensor(t, t->_dimensions, t->requires_grad);

	// Apply Sigmoid -> r = 1 / (1 + e^(-t))
	MathUtils::vector_sigmoid(t->_data, result->_data);

	// Connect result tensor to autograd graph if needed
	if (!inference)
	{
		result->_parents.push_back(t);

		// Gradient calculation function -> dr/dt = Sigmoid(t) * (1 - Sigmoid(t))
		result->_backward = [t, result]()
			{
				if (!t->requires_grad) return;

				thread_local AlignedFloatVector scratch1;

				const int element_count = result->element_count();
				scratch1.resize(element_count);
				MathUtils::vector_sub(1.0f, result->_data, scratch1);
				MathUtils::vector_mul(scratch1, result->_data);
				MathUtils::vector_fmadd(t->_grad, scratch1, result->_grad);
			};
	}

	return result;
}

std::shared_ptr<Tensor> Tensor::tanh(const std::shared_ptr<Tensor>& t)
{
	auto result = get_result_tensor(t, t->_dimensions, t->requires_grad);

	// Apply Tanh -> r = (e^(2t) - 1) / (e^(2t) + 1)
	MathUtils::vector_tanh(t->_data, result->_data);

	// Connect result tensor to autograd graph if needed
	if (!inference)
	{
		result->_parents.push_back(t);

		// Gradient calculation function -> dr/dt = 1 - Tanh^2(t)
		result->_backward = [t, result]()
			{
				if (!t->requires_grad) return;

				thread_local AlignedFloatVector scratch1;

				const int element_count = result->element_count();
				scratch1.resize(element_count);
				MathUtils::vector_sq(result->_data, scratch1);
				MathUtils::vector_sub(1.0f, scratch1);
				MathUtils::vector_fmadd(t->_grad, scratch1, result->_grad);
			};
	}

	return result;
}

std::shared_ptr<Tensor> Tensor::softmax(const std::shared_ptr<Tensor>& t)
{
	auto result = get_result_tensor(t, t->_dimensions, t->requires_grad);

	const size_t classes = t->_dimensions.back();
	const size_t batches = t->element_count() / classes;

	// Apply Softmax per batch -> Softmax(z_i) = e^z_i / (sum(j = 1 to n)[e^z_j])
	for (size_t b = 0; b < batches; ++b)
	{
		const size_t offset = b * classes;

		const float* const __restrict p_t = t->_data.data() + offset;
		float* const __restrict p_r = result->_data.data() + offset;

		const float max = MathUtils::vector_max(p_t, classes);

		const __m256 reg_max = _mm256_set1_ps(max);
		__m256 acc0 = _mm256_setzero_ps();
		__m256 acc1 = _mm256_setzero_ps();
		__m256 acc2 = _mm256_setzero_ps();
		__m256 acc3 = _mm256_setzero_ps();

		// AVX2 SIMD vectorize result calculation
		size_t i = 0;
		for (; i + 32 <= classes; i += 32)
		{
			const __m256 exp0 = _mm256_exp_ps(_mm256_sub_ps(_mm256_load_ps(&p_t[i]), reg_max));
			const __m256 exp1 = _mm256_exp_ps(_mm256_sub_ps(_mm256_load_ps(&p_t[i + 8]), reg_max));
			const __m256 exp2 = _mm256_exp_ps(_mm256_sub_ps(_mm256_load_ps(&p_t[i + 16]), reg_max));
			const __m256 exp3 = _mm256_exp_ps(_mm256_sub_ps(_mm256_load_ps(&p_t[i + 24]), reg_max));

			_mm256_store_ps(&p_r[i], exp0);
			_mm256_store_ps(&p_r[i + 8], exp1);
			_mm256_store_ps(&p_r[i + 16], exp2);
			_mm256_store_ps(&p_r[i + 24], exp3);

			acc0 = _mm256_add_ps(acc0, exp0);
			acc1 = _mm256_add_ps(acc1, exp1);
			acc2 = _mm256_add_ps(acc2, exp2);
			acc3 = _mm256_add_ps(acc3, exp3);
		}

		acc0 = _mm256_add_ps(acc0, acc1);
		acc2 = _mm256_add_ps(acc2, acc3);
		acc0 = _mm256_add_ps(acc0, acc2);

		for (; i + 8 <= classes; i += 8)
		{
			const __m256 exp = _mm256_exp_ps(_mm256_sub_ps(_mm256_load_ps(&p_t[i]), reg_max));
			_mm256_store_ps(&p_r[i], exp);
			acc0 = _mm256_add_ps(acc0, exp);
		}

		float sum = MathUtils::sum_m256(acc0);

		for (; i < classes; ++i)
		{
			const float exp = std::exp(p_t[i] - max);
			p_r[i] = exp;
			sum += exp;
		}

		// Normalize result values
		MathUtils::vector_div(p_r, sum, classes);
	}

	// Connect result tensor to autograd graph if needed
	if (!inference)
	{
		result->_parents.push_back(t);

		// Gradient calculation function -> grad_t_i = r_i * (grad_r_i - sum_j(grad_r_j * r_j))
		result->_backward = [t, result, batches, classes]()
			{
				if (!t->requires_grad) return;

				thread_local AlignedFloatVector scratch1;

				scratch1.resize(classes);

				float* const __restrict p_tg = t->_grad.data();
				const float* const __restrict p_rv = result->_data.data();
				const float* const __restrict p_rg = result->_grad.data();

				// Compute gradients per batch
				for (size_t b = 0; b < batches; ++b)
				{
					const size_t offset = b * classes;

					const float dot = MathUtils::vector_dot(p_rg + offset, p_rv + offset, classes);
					MathUtils::vector_sub(p_rg + offset, dot, scratch1.data(), classes);
					MathUtils::vector_fmadd(p_tg + offset, p_rv + offset, scratch1.data(), classes);
				}
			};
	}

	return result;
}

std::shared_ptr<Tensor> Tensor::linear(const std::shared_ptr<Tensor>& t)
{
	auto result = get_result_tensor(t, t->_dimensions, t->requires_grad);

	// Copy exact data - linear function
	std::copy(t->_data.begin(), t->_data.end(), result->_data.begin());

	// Connect result tensor to autograd graph if needed
	if (!inference)
	{
		result->_parents.push_back(t);

		// Gradient calculation function -> dr/dt = 1
		result->_backward = [t, result]()
			{
				MathUtils::vector_add(t->_grad, result->_grad);
			};
	}

	return result;
}